design-proposals/kliup-29-explicit-keys.md

```diff
@@ -0,0 +1,154 @@
+Tables require a PRIMARY KEY. Please define the PRIMARY KEY.
+```
+
+Statements with a `PRIMARY KEY` column will continue to work as before.
+
+Where schema inference is used, the error message will include an example of how to define a partial
+schema to add the primary key:
+
+```ql
+ksql> CREATE TABLE INPUT WITH (value_foramt='Avro', ...);
+Tables require a PRIMARY KEY. Please define the PRIMARY KEY.
+You can define just the primary key and still load the value columns from the Schema registry, for example:
+  CREATE TABLE INPUT (ID INT PRIMARY KEY) WITH (value_foramt='Avro', ...);
+```
+
+### CREATE STREAM changes
+
+A `CREATE STREAM` statement that does not define a `KEY` column will now result in a stream with no
+data being read from the Kafka message's key.
+
+```sql
+CREATE STREAM INPUT (A INT, B STRING) WITH (...);
+-- Results in INPUT with schema: A INT, B STRING, i.e. no key column.
+```
+
+Statements with a `KEY` column will continue to work as before.
+
+## Design
+
+This is mainly a syntax only change, as detailed about. 
+
+Streams with out a key column will work just like any other stream. However, `GROUP BY`, `PARTITION BY`
+and `JOIN`s will _always_ result in a repartition, as the grouping, partitioning or joining expression
+can never be the key column.
+
+## Test plan
+
+Ensure coverage of key-less streams in QTT tests, especially for `GROUP BY`, `PARTITION BY`, and `JOIN`s.
+
+## LOEs and Delivery Milestones
+
+Single deliverable, with low loe (prototype already working).
+
+## Documentation Updates
+
+Suitable doc updates for the `CREATE TABLE` and `CREATE STREAM` statements will be done as part of the KLIP.
+
+Plus updates to the rests of the ksqlDB docs, Kafka tutorials microsite and the Examples repo will be done
+in tandem with other syntax changes.
+
+Release notes to call out this change in behaviour.
+
+## Compatibility Implications
+
+CREATE Statements submitted on previous versions of ksqlDB will continue to work as expected.
+
+Users submitting previously written statements may see `CREATE TABLE` statements that previously ran,
+now fail, and see `CREATE STREAM` statements create streams without a `ROwKEY STRING KEY` column. 
+
+Users receiving an error when their `CREATE TABLE` statements fail will need to update their statements
+to include a suitable `PRIMARY KEY` column.  Where the statement already contains the column set, the 
+addition of the `PRIMARY KEY` column should be simple.  However, users may be more confused when the 
+statement is making use of schema inference, i.e. loading the value columns from the Schema Registry,
+for example:
+
+```sql
+-- existing create statement that loads the value columns from the Schema Registry: 
+CREATE TABLE OUTPUT WITH (value_format='Avro', kafka_topic='input');
+```
+
+As ksqlDB does not _yet_ support loading the key schema from the Schema Registry the user must now
+supply the `PRIMARY KEY` in a _partial schema_. (_Partial schema_ support was added in v0.9):
```

Comment:
Just to clarify: After key schemas are integrated with SR (and after we fully support structured keys), a user will be able to do
```
CREATE TABLE myTable WITH (key_format='Avro', value_format='Avro', kafka_topic='input');
```
And the all columns from the key-avro-schema will be the PK automatically?

Reply:
Yup